command/agent/cache/lease_cache.go

```diff
@@ -0,0 +1,743 @@
+		return fmt.Errorf("unable to convert %v to type string", clearValue)
+	}
+
+	switch clearType {
+	case "request_path":
+		// For this particular case, we need to ensure that there are 2 provided
+		// indexers for the proper lookup.
+		if len(clearValues) != 2 {
+			return fmt.Errorf("clearing cache by request path requires 2 indexers, got %d", len(clearValues))
+		}
+
+		// The first value provided for this case will be the namespace, but if it's
+		// an empty value we need to overwrite it with "root/" to ensure proper
+		// cache lookup.
+		if clearValues[0].(string) == "" {
+			clearValues[0] = "root/"
+		}
+
+		// Find all the cached entries which has the given request path and
+		// cancel the contexts of all the respective renewers
+		indexes, err := c.db.GetByPrefix(clearType, clearValues...)
+		if err != nil {
+			return err
+		}
+		for _, index := range indexes {
+			index.RenewCtxInfo.CancelFunc()
+		}
+
+	case "token":
+		if clearValue == "" {
+			return nil
+		}
+		// Get the context for the given token and cancel its context
+		tokenCtxInfo := c.tokenContexts[clearValue]
+		if tokenCtxInfo == nil {
+			return nil
+		}
+
+		tokenCtxInfo.CancelFunc()
+
+		// Remove the cancelled context from the map
+		delete(c.tokenContexts, clearValue)
+
+	case "token_accessor", "lease":
+		// Get the cached index and cancel the corresponding renewer context
+		index, err := c.db.Get(clearType, clearValue)
+		if err != nil {
+			return err
+		}
+		if index == nil {
+			return nil
+		}
+		index.RenewCtxInfo.CancelFunc()
+
+	case "all":
+		// Cancel the base context which triggers all the goroutines to
+		// stop and evict entries from cache.
+		c.baseCtxInfo.CancelFunc()
+
+		// Reset the base context
+		baseCtx, baseCancel := context.WithCancel(ctx)
+		c.baseCtxInfo = &ContextInfo{
+			Ctx:        baseCtx,
+			CancelFunc: baseCancel,
+		}
+
+		// Reset the memdb instance
+		if err := c.db.Flush(); err != nil {
+			return err
+		}
+
+	default:
+		return errInvalidType
+	}
+
+	c.logger.Debug("successfully cleared matching cache entries")
+
+	return nil
+}
+
+// handleRevocationRequest checks whether the originating request is a
+// revocation request, and if so perform applicable cache cleanups.
+// Returns true is this is a revocation request.
+func (c *LeaseCache) handleRevocationRequest(ctx context.Context, req *SendRequest, resp *SendResponse) (bool, error) {
+	// Lease and token revocations return 204's on success. Fast-path if that's
+	// not the case.
+	if resp.Response.StatusCode != http.StatusNoContent {
+		return false, nil
+	}
+
+	namespace, path := deriveNamespaceAndRevocationPath(req)
+
+	switch {
+	case path == vaultPathTokenRevoke:
+		// Get the token from the request body
+		jsonBody := map[string]interface{}{}
+		if err := json.Unmarshal(req.RequestBody, &jsonBody); err != nil {
+			return false, err
+		}
+		token, ok := jsonBody["token"]
+		if !ok {
+			return false, fmt.Errorf("failed to get token from request body")
+		}
```

Comment:
For safety can we verify token is a string here with an `, ok`?

Reply:
Fixed.